### PR TITLE
allow integers instead of layer names in embedding extraction

### PR DESCRIPTION
**Pull request by nkundiushuti**
Increase usability of embedding extraction, currently this is done by specifying the exact layer name which is cumbersome. We allow to add layers as integer indices:

- Add 0-based and negative integer indices to `register_hooks_for_layers` for simpler layer selection (`avex/models/base_model.py`)
- Allow `ProbeConfig.target_layers` to accept `int` indices and forbid `bool` coercion (`avex/configs.py`)
- Resolve integer layer identifiers to concrete layer names in embedding dataset creation (`avex/evaluation/embedding_manager.py`)
- Make embedding output collection deterministic by preserving hook registration order (`avex/models/base_model.py`)
- Add clear guardrails for TF/TFLite models that cannot support PyTorch hook-based intermediate layers (`avex/models/perch.py`, `avex/models/surfperch.py`, `avex/models/birdnet.py`)
- Document BirdMAE intermediate-layer limitation in the same layer-selection story (`avex/models/birdmae.py`)
- Update docs and examples to demonstrate index-based layer selection (`README.md`, `docs/embedding_extraction.md`, `docs/api_probes.md`, `docs/examples.md`, `examples/05_embedding_extraction.py`, `examples/07_probe_training_and_inference.py`, `examples/08_probe_comparison.py`)
- Propagate `str | int` layer identifiers into evaluation filename generation (`avex/run_evaluate.py`)
- Add unit tests covering integer/negative layer indices and `ProbeConfig` validation (`tests/unittests/test_base_model_all_layers.py`, `tests/unittests/test_api_probes.py`)
- Keep model wrappers compatible with the updated hook-based extraction flow (`avex/models/beats_model.py`, `avex/models/efficientnet.py`, `avex/models/eat_hf.py`)

### CODE REVIEW

File: avex/models/base_model.py [hunk 1]
```diff
@@ -49,6 +49,42 @@ def _discover_linear_layers(self) -> None:
...
+        Returns
+        -------
+        list[str]
+            Discoverable layer names in index order.
+        """
+        if len(self._layer_names) == 0:
+            self._discover_linear_layers()
+        return self._layer_names.copy()
+
+    def get_model_num_layers(self) -> int:
```
Comment (GaganNarula) [inline avex/models/base_model.py]: seems we're adding a lot more methods than necessary.. this one here just calls `len()` .. couldn't the user do that ?

File: avex/models/base_model.py [hunk 2]
```diff
@@ -73,7 +109,7 @@ def hook_fn(
 
         return hook_fn
 
-    def register_hooks_for_layers(self, layer_names: List[str]) -> List[str]:
+    def register_hooks_for_layers(self, layer_names: List[Union[str, int]]) -> List[str]:
```
Comment (GaganNarula) [inline avex/models/base_model.py]: `layer_names` doesnt work as a param name for an `int` type. Would have been better to use `layer_id` but i guess this would break backwards compat

File: examples/08_probe_comparison.py [hunk 3]
```diff
@@ -56,7 +56,7 @@ def compute_embeddings(
     model: nn.Module,
     dataloader: DataLoader,
     device: str,
-    target_layers: list[str] = None,
+    target_layers: list[str | int] = None,
```
[the remark this answers is not in the document]
Comment (nkundiushuti) [reply examples/08_probe_comparison.py]: Good catch — updated `compute_embeddings` to use `target_layers: list[str | int] | None = None` so the default matches the annotation.

File: docs/examples.md [hunk 4]
```diff
@@ -12,6 +12,7 @@ The `examples/` directory contains comprehensive examples demonstrating various
 | `05_embedding_extraction.py` | Feature extraction with `return_features_only=True` (unpooled features) |
 | `06_classifier_head_loading.py` | Understanding classifier head behavior with different `num_classes` settings |
 | `colab_sl_beats_demo.ipynb` | Google Colab demo for the sl-beats model |
+| `09_multi_layer_embedding_extraction.py` | Index-based multi-layer embedding extraction (`register_hooks_for_layers([0, -1])`) |
```
[the remark this answers is not in the document]
Comment (nkundiushuti) [reply docs/examples.md]: Removed the dangling `09_multi_layer_embedding_extraction.py` entry. The examples table now documents index-based extraction on `08_probe_comparison.py` instead.

File: avex/evaluation/embedding_manager.py [hunk 5]
```diff
@@ -250,6 +251,29 @@ def get_dataset(
         first_batch = next(iter(dataloader))
         base_model.eval()
 
+        # If layer_names include integer indices, resolve them against this loaded
+        # model once so downstream logic (filtering by keys, filenames, etc.) sees
+        # concrete string names.
+        if any(isinstance(x, int) for x in self.layer_names):
+            if hasattr(base_model, "get_model_layers"):
+                discovered = base_model.get_model_layers()  # type: ignore[attr-defined]
```
[the remark this answers is not in the document]
Comment (nkundiushuti) [reply avex/evaluation/embedding_manager.py]: Load-only mode now resolves integer indices against the HDF5 `layer_names` metadata before building `HDF5EmbeddingDataset` or filtering in-memory embeddings. If metadata is missing, we raise a clear error instead of passing ints through as HDF5 keys.

File: avex/models/base_model.py (hunk 1 above)
Comment (nkundiushuti) [reply avex/models/base_model.py]: dropped `get_model_num_layers()`

File: avex/models/base_model.py (hunk 2 above)
Comment (nkundiushuti) [reply avex/models/base_model.py]: agreed, I would go for target_layers which doesn't imply a type and this is what we used for the model classes

File: avex/models/base_model.py [hunk 6]
```diff
@@ -49,6 +49,32 @@ def _discover_linear_layers(self) -> None:
...
+        The layer order defines the indexing used by `target_layers=[...]` in
+        `extract_embeddings` (0-based, supports negative indices).
+
+        Returns
+        -------
+        list[str]
+            Discoverable layer names in index order.
+        """
+        if len(self._layer_names) == 0:
+            self._discover_linear_layers()
```
Comment (GaganNarula) [inline avex/models/base_model.py]: is this the only place where `self._layer_names` is set ? so are linear layers the only kind of layers ? what if a model has conv layers or attention layers ?

Comment (GaganNarula) [reply avex/models/base_model.py]: ok i looked through the code and it looks like each model overrides this so its no just linear layers. More comments from Claude review (/review didnt catch it on its own, had to guide it towards this):

Naming/dead-code: _discover_linear_layers is a misnomer, and _discover_embedding_layers is dead.
  
  self._layer_names isn't only set here, and it isn't only linear layers. Every backbone overrides _discover_linear_layers, and those
  overrides already match non-Linear modules (e.g. BEATs picks post_extract_proj + each block's .fc2). The base version's "only
  nn.Linear" behavior is just a fallback — so the live behavior is fine, but the name is misleading.

  Two cleanups worth doing while we're in here:

  1. _discover_embedding_layers has zero callers anywhere in avex/ — the live discovery path is _discover_linear_layers (called from
  get_model_layers, register_hooks_for_layers, and registry.list_model_layers). The parallel _discover_embedding_layers in all the
  subclasses is dead code (some with commented-out bodies). Suggest deleting it.
  2. After deleting it, rename _discover_linear_layers → _discover_embedding_layers so the name reflects that it discovers embedding
  taps (conv/attention/proj), not just nn.Linear.

  Risk this leaves: if a new backbone forgets to override the discovery method, the base falls back to Linear-only (or []), and integer
   indexing then silently selects wrong/no layers. Consider having get_model_layers() raise if _layer_names is empty so that fails
  loudly instead.
  
  (Also the if len(self._layer_names) == 0: guard on line 63 duplicates the identical check inside the method — can just call it
  unconditionally.)

**Review by GaganNarula — APPROVED** [review]
approving this, but one thing we might want to revisit is that when a user just wants to discover the layers / architecture of the model, for e.g. to choose which layers to use for embedding extraction or probing, they have to download the model weights so that `list_model_layers` can return an ordered list.. this shouldn't be necessary! worth making an issue
